### PR TITLE
WalletConnect v2 #3413

### LINKED ISSUE #3413: WalletConnect v2
* Docs: https://docs.walletconnect.com
* SDK
    * Swift: https://github.com/WalletConnect-Labs/WalletConnectSwiftV2
        * Doc mentions requires minimum deployment iOS 13 (but doesn't seem so in practice?)
    * Kotlin: https://github.com/WalletConnect-Labs/WalletConnectKotlinV2
* Test sites
    Test wallet site
       https://react-wallet.walletconnect.com
    Test dapp site
        https://react-app.walletconnect.com

Observations
---
* Different version number in the wc URL
    * v2 vs v1 example (@2 vs @1)
    * wc:ea6410bc3ea20864f4331095335e0d3a23f5e70820f00f8edae9ec14eeb33917@2?controller=false&publicKey=35af052875ac113c6efab51934c8fa7967990345b6dfeb2a0a8b964693350c32&relay=%7B%22protocol%22%3A%22waku%22%7D and wc:2fa9b5c0-37f9-498f-b975-7874a6cf74b5@1?bridge=https%3A%2F%2Fs.bridge.walletconnect.org&key=6d9e4064d39e574dcdb0218ab31b2d2bb2a60ca15ebf42d0ead9d2a327de396f
* Allow requesting a set of chains
* Wallet no longer controls the chain ID of the chain, but session tracks multiple chain IDs
* Dapp requests permission to use a set of JSON-RPC methods (and whether they are supported)
* Session expiry, default to 7 days
* What are "pairings"? Decouples session from peer to peer pair
* Each session gets (after requesting and granted access) access to multiple wallets/accounts and chains

Wallet
---
* List all active sessions
* User can expand each session to check which account(s) and chain(s)
* Disconnect sessions
* Grant permissions for requested (and supported) JSON-RPC methods dapp needs

> Dapps should request upfront as many chains as possible from the wallet to expose in order to allow the user to control which chain and/or account is using directly from the dapp.
    https://docs.walletconnect.com/protocol/ux-considerations
    Doesn't sound like a good design choice

> Account decoupling
    This more relevant for Wallets where some multi-account wallets have the concept of an "active" account on their interface. This should be completely decoupled from the session state on WalletConnect 2.0. A user should be able to expose and sign methods targetted at an account that is different than the active one in the wallet.
        https://docs.walletconnect.com/protocol/ux-considerations
        We can be limited to 1 account at a time (because we have the concept of an active wallet), for now?

### PR DESCRIPTION
**Pull request by oa-s**
Closes #3413

### CODE REVIEW

**Comment by oa-s** [comment]
It seems like enabling rosetta mode for xCode helps to resolve building the proj for simulator

**Comment by hboon** [comment]
Yes, it would, but I think it kills the performance gains from running Xcode on M1. Do you experience it? Is there a better way?

**Comment by oa-s** [comment]
> Yes, it would, but I think it kills the performance gains from running Xcode on M1. Do you experience it? Is there a better way?

will try to find out it

**Comment by oa-s** [comment]
@hboon @eviltofu guys, can u check if this PR is building for you? Because it removes changes introduced in https://github.com/AlphaWallet/alpha-wallet-ios/pull/3533, and need to make sure that everything is ok.

**Comment by hboon** [comment]
@oa-s It doesn't work if I clone the repo again, and

1. run:

```
$ bundle && bundle exec pod install
```

2. Xcode > File > Packages > Resolve Package Version (and Xcode build logs suggested WalletConnectSwiftv2, MailChimp-SDK-iOS and CryptoSwift being resolved)

3. and build for both simulator and device:

> Analyze workspace
> Create build description
> Build description signature: 17f6464df8b63a579b1e059d17c77369
> Build description path: /Users/hboon/Library/Developer/Xcode/DerivedData/AlphaWallet-ceryaordvcedbeacqvjohckfveup/Build/Intermediates.noindex/XCBuildData/17f6464df8b63a579b1e059d17c77369-desc.xcbuild
> 
> 
> Build target AlphaWallet of project AlphaWallet with configuration Debug
> warning: no rule to process file '/Users/hboon/Desktop/alpha-wallet-ios/AlphaWallet/Core/Enjin/EnjinBalances.graphql' of type 'text' for architecture 'arm64' (in target 'AlphaWallet' from project 'AlphaWallet')
> /Users/hboon/Desktop/alpha-wallet-ios/AlphaWallet.xcodeproj: error: Missing package product 'MailchimpSDK' (in target 'AlphaWallet' from project 'AlphaWallet')

Using Xcode 13.2.1. It seemed to work with my original copy of the repo even after cleaning.

**Comment by eviltofu** [comment]
MacBook Air (M1, 2020)
macOS 12.1
Xcode 13.2.1
Target: iPhone 12/iOS 14.5 & iPhone12/iOS15.2

This branch set up via:
```
make clean && make bootstrap
```

Then I ran the unit tests and they all passed.

**Comment by oa-s** [comment]
> bundle && bundle exec pod install

@hboon i followed the same steps, i am able to build it, compiles without error `error: Missing package product 'MailchimpSDK' (in target 'AlphaWallet' from project 'AlphaWallet')`

**Comment by hboon** [comment]
> MacBook Air (M1, 2020) macOS 12.1 Xcode 13.2.1 Target: iPhone 12/iOS 14.5 & iPhone12/iOS15.2
> 
> This branch set up via:
> 
> ```
> make clean && make bootstrap
> ```
> 
> Then I ran the unit tests and they all passed.

Both simulator and device? If there's a problem, it's more likely on the simulator than device.

**Comment by oa-s** [comment]
all tests are passing for simulator too

**Review by hboon — APPROVED** [review]
Builds for me too. Too big to go through. If you are fairly confident WalletConnect v1 works, let's get this merged.

Can you summarize what charged to make it build? I also noticed that we are using the Cocoapod version of MailChimp now.

**Comment by oa-s** [comment]
Fortunally we was able to use pod for MailChimp, because of setting `      config.build_settings["ARCHS[sdk=iphonesimulator*]"] = "x86_64"` for project brokes building WalletConnect, that is why we moved it to pod and configured it there. as i found in example project of WalletConnect v2 works fine.
